docs/architecture/cloud-native/rest-grpc.md

```diff
@@ -1,69 +1,347 @@
-- Point-to-point real-time communication. gRPC has excellent support for bi-directional streaming. gRPC services can push messages in real time without polling.
-- Polyglot environments – gRPC tooling supports most popular development languages, making it a good choice for multi-language environments.
-- Network constrained environments – gRPC messages are serialized with Protobuf, a lightweight message format. A gRPC message is always smaller than an equivalent JSON message.
+Because the Shopping Basket microservices exposes both a RESTful API and gRPC, we need to open two different endpoints to manage this traffic. Figure 4-29 shows the Program.cs file for the microservice.
 
-At the time of writing of this book, most browsers have limited support for gRPC. gRPC heavily uses HTTP/2 features and no browser provides the level of control required over web requests to support a gRPC client. gRPC is typically used for internal microservice to microservice communication. Figure 4-22 shows a simple, but common usage pattern.
+```csharp
+    private static IWebHost BuildWebHost(IConfiguration configuration, string[] args) =>
+        WebHost.CreateDefaultBuilder(args)
+            .CaptureStartupErrors(false)
+            .ConfigureKestrel(options =>
+            {
+                var ports = GetDefinedPorts(configuration);
+                // Endpoint for RESTful calls
+                options.Listen(IPAddress.Any, ports.httpPort, listenOptions =>
+                {
+                    listenOptions.Protocols = HttpProtocols.Http1AndHttp2;
+                });
 
-![gRPC Usage Patterns](./media/grpc-usage.png)
+                // Endpoint for gRPC calls
+                options.Listen(IPAddress.Any, ports.grpcPort, listenOptions =>
+                {
+                    listenOptions.Protocols = HttpProtocols.Http2;
+                });
+            })
+```
 
-**Figure 4-22**. gRPC usage patterns
+Note in the previous figure how two endpoints are opened. One listens for HTTP traffic for the RESTful endpoints and the other for gRPC calls. Importantly, the gRPC endpoint specifies the HTTP/2 protocol which is required for gRPC communication.
 
-Note in the previous figure how front-end traffic is invoked with HTTP while back-end microservice to microservice uses gRPC.
+## Looking ahead
 
-Looking ahead, gRPC could play a major role in dethroning the dominance of REST for cloud-native systems. The performance benefits and ease of development are too good to pass up. However, make no mistake, REST will still be around for a long time. It still excels for publicly exposed APIs and for backward compatibility reasons.
+Looking ahead, gRPC could well play a major role in dethroning the dominance of REST for cloud-native systems. The performance benefits and ease of development are compelling. However, REST will still most likely be around for a long time. It excels for publicly exposed APIs and for backward compatibility reasons.
```

Comment:
The use of a word like dethroning is unnecessary. We should be subtle in our prediction statements :)

Reply:
Remove "dethroning" and softened prediction.